File: Python3/프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_map = {}
    
    # map에 삽입 
    for phone in phone_book: 
        phone_map[phone] = 1 
    
    # phone_book 리스트 순회
    for phone in phone_book: 
        arr = "" 
        
        # string에 전화번호 삽입
        for num in phone: 
            arr += num
            
            # 현재 번호와 같지 않고 map 안에 있으면 False return
            if arr in phone_map and arr != phone:       
                return False 
                
    return True

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <unordered_set>
#include <string>
#include <vector>

using namespace std;

bool solution(vector<string> phone_book) {
    unordered_set<string> us(phone_book.begin(), phone_book.end());
    
    for(const auto& s: us) {
        for(int i = 1; i < s.size(); i++) {
            if(us.find(s.substr(0, i)) != us.end()) {
                return false;
            }
        }
    }
    
    return true;
}